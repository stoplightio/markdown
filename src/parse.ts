import remarkFrontmatter from 'remark-frontmatter';
import remarkGFM from 'remark-gfm';
import remarkParse, { RemarkParseOptions } from 'remark-parse';
import unified from 'unified';
import type { VFileCompatible } from 'vfile';

import { MDAST } from './ast-types';
import {
  blockquoteMdast2Hast,
  inlineCodeMdast2Hast,
  inlineImages,
  resolveCodeBlocks,
  Resolver,
  slug,
  smdAnnotations,
  smdCode,
  unwrapImages,
} from './plugins/run';
import { replaceThirdPartyBlocks } from './replaceThirdPartyBlocks';

export type ParseSettings = RemarkParseOptions;

export type ParseOptions = {
  remarkPlugins?: unified.PluggableList<unified.Settings>;
  settings?: ParseSettings;
};

export type AsyncParseOptions = {
  remarkPlugins?: unified.PluggableList<unified.Settings>;
  settings?: ParseSettings & {
    resolver?: Resolver;
  };
};

export const remarkParsePreset: unified.Preset<ParseSettings> = {
  plugins: [
    [remarkFrontmatter, ['yaml']],
    remarkGFM,
    slug,
    unwrapImages,
    smdAnnotations,
    smdCode,
    inlineImages,
    inlineCodeMdast2Hast,
    blockquoteMdast2Hast,
  ],
  settings: {},
};

const defaultProcessor = unified().use(remarkParse).use(remarkParsePreset);

export const parse = (
  input: VFileCompatible,
  opts: Partial<ParseOptions> = {},
  processor: unified.Processor = defaultProcessor,
): MDAST.Root => {
  const markdown = replaceThirdPartyBlocks(input);

  const processorInstance = processor()
    .data('settings', Object.assign({}, remarkParsePreset.settings, opts.settings))
    .use(opts.remarkPlugins || []);

  // return the parsed remark ast
  return processorInstance.runSync(processorInstance.parse(markdown)) as MDAST.Root;
};

export const parseAsync = (
  input: VFileCompatible,
  opts: Partial<AsyncParseOptions> = {},
  processor: unified.Processor = defaultProcessor,
): Promise<MDAST.Root> => {
  const markdown = replaceThirdPartyBlocks(input);

  const processorInstance = processor()
    .data('settings', Object.assign({}, remarkParsePreset.settings, opts.settings))
    .use(resolveCodeBlocks, { resolver: opts.settings?.resolver })
    .use(opts.remarkPlugins || []);

  // return the parsed remark ast
  return processorInstance.run(processorInstance.parse(markdown)) as Promise<MDAST.Root>;
};
