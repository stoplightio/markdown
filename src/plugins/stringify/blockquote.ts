import { safeStringify } from '@stoplight/yaml';
import { Handler } from 'mdast-util-to-markdown';
// @ts-expect-error
import blockquote from 'mdast-util-to-markdown/lib/handle/blockquote';

export const blockquoteHandler: Handler = function (node, _, context) {
  const annotations = (node.data?.hProperties || {}) as any;

  const value = blockquote(node, _, context);

  if (Object.keys(annotations).length) {
    return `<!-- ${safeStringify(annotations, { skipInvalid: true }).trim()} -->

${value}`;
  } else {
    return value;
  }
};
