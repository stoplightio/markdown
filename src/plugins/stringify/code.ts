import * as Yaml from '@stoplight/yaml';
import { Handler } from 'mdast-util-to-markdown';
// @ts-expect-error
import code from 'mdast-util-to-markdown/lib/handle/code';

const { safeStringify } = Yaml;

export const codeHandler: Handler = function (node, _, context) {
  const { lang, meta: _meta, ...annotations } = (node.data?.hProperties || {}) as any;

  // if node has resolved prop we want to replace value with resolved value when stringifying
  if (node.resolved) {
    node.value =
      node.lang === 'json' ? JSON.stringify(node.resolved, null, 2) : safeStringify(node.resolved, { indent: 2 });
  }

  const metaProps = computeMetaProps(annotations);
  if (metaProps.length) {
    node.meta = metaProps.join(' ');
  }

  return code(node, _, context);
};

function computeMetaProps(annotations: object) {
  const metaProps = [];
  if (Object.keys(annotations).length) {
    for (const key in annotations) {
      const annotationVal = annotations[key];

      if (typeof annotationVal === 'boolean' || annotationVal === 'true' || annotationVal === 'false') {
        if (annotationVal || annotationVal === 'true') {
          metaProps.push(key);
        }

        // don't add falsey val to meta string
        continue;
      } else if (key === 'type') {
        // this is the only old val we support
        if (annotationVal === 'json_schema') {
          // camelCase to be consistent with rest of annotation props
          metaProps.push('jsonSchema');
        }
      } else if (key === 'highlightLines') {
        // handle deprecated way of adding highlightLines via array annotation
        if (Array.isArray(annotationVal)) {
          const rangeVals = [];
          for (const val of annotationVal) {
            if (Array.isArray(val)) {
              rangeVals.push(`${val[0]}-${val[1]}`);
            } else {
              rangeVals.push(val);
            }
          }
          if (rangeVals.length) {
            metaProps.push(`{${rangeVals.join(',')}}`);
          }
        } else {
          // else we're dealing with the new {1,3} style highlightLines
          metaProps.push(`{${annotationVal}}`);
        }
      } else {
        metaProps.push(`${key}="${annotationVal}"`);
      }
    }
  }

  return metaProps;
}
