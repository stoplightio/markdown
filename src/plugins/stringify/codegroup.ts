import { Handler } from 'mdast-util-to-markdown';
// @ts-expect-error
import flow from 'mdast-util-to-markdown/lib/util/container-flow';

export const codeGroupHandler: Handler = (node, _, context) => {
  const exit = context.enter('codegroup');
  const value = flow(node, context);
  exit();
  return value;
};
