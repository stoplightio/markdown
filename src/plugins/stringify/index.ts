export * from './blockquote';
export * from './code';
export * from './tabs';
