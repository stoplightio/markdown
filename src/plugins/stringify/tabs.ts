import * as Yaml from '@stoplight/yaml';
import { Handler } from 'mdast-util-to-markdown';
// @ts-expect-error
import flow from 'mdast-util-to-markdown/lib/util/container-flow';

const { safeStringify } = Yaml;

export const tabsHandler: Handler = function (node, _, context) {
  const exit = context.enter('tabs');
  const value = flow(node, context);
  exit();
  return `${value}

<!-- type: tab-end -->`;
};

export const tabHandler: Handler = function (node, _, context) {
  const exit = context.enter('tab');

  const { type, ...annotations } = (node.data?.hProperties || {}) as any;

  const value = flow(node, context);

  exit();

  return `<!--
type: tab
${safeStringify(annotations, { skipInvalid: true }).trim()}
-->

${value}`;
};
