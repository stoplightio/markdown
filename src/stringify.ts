import unified, { Processor } from 'unified';
import remarkStringify, { StringifyOpts } from 'remark-stringify';
import remarkParse from 'remark-parse';
import { INode as Node } from './ast-types/unist';

const defaultOpts: StringifyOpts = {
  commonmark: true,
  gfm: true,
};

const defaultProcessor = unified()
  .use(remarkParse)
  .use(remarkStringify);

export const stringify = (tree: Node, opts: StringifyOpts = defaultOpts, processor: Processor = defaultProcessor) => {
  return processor
    .data('settings', defaultOpts)
    .stringify(tree);
};
