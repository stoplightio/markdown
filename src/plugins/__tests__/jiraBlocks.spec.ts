import fs from 'fs';
import { join } from 'path';
import { parse } from '../../parse';
import { stringify } from '../../stringify';

describe('Jira blocks plugin', () => {
  it('should parse simple jira block', () => {
    expect(
      parse(`[block:image]
{}
[/block]`),
    ).toStrictEqual(
      expect.objectContaining({
        children: [
          {
            code: 'image',
            position: expect.objectContaining({
              end: {
                column: 9, // markdown lines are columns is 1-based
                line: 3,
                offset: 25,
              },
              start: {
                column: 1,
                line: 1,
                offset: 0,
              },
            }),
            type: 'jira',
            value: '{}',
          },
        ],
        type: 'root',
      }),
    );
  });

  it('should parse more complex document correctly', () => {
    expect(parse(fs.readFileSync(join(__dirname, '__fixtures__/jira.md'), 'utf8'))).toMatchSnapshot();
  });

  it('should not recognize blocks contained in fenced code', () => {
    expect(
      parse(`\`\`\`
[block:image]
{}
[/block]
\`\`\``),
    ).toStrictEqual(
      expect.objectContaining({
        type: 'root',
        children: [
          expect.objectContaining({
            type: 'code',
            lang: null,
            meta: null,
            value: '[block:image]\n{}\n[/block]',
          }),
        ],
      }),
    );
  });

  describe('should not recognize invalid blocks', () => {
    it('no new-line', () => {
      expect(
        parse(`[block:image] {}
[/block]`),
      ).toStrictEqual(
        expect.objectContaining({
          children: expect.not.arrayContaining([
            expect.objectContaining({
              type: 'jira',
            }),
          ]),
        }),
      );
    });

    it('no code', () => {
      expect(
        parse(`[block]
{}
[/block]`),
      ).toStrictEqual(
        expect.objectContaining({
          children: expect.not.arrayContaining([
            expect.objectContaining({
              type: 'jira',
            }),
          ]),
        }),
      );
    });

    it('no ending', () => {
      expect(
        parse(`[block:code]
{}
`),
      ).toStrictEqual(
        expect.objectContaining({
          children: expect.not.arrayContaining([
            expect.objectContaining({
              type: 'jira',
            }),
          ]),
        }),
      );
    });
  });

  it('should stringify markdown containing jira blocks', () => {
    const document = fs.readFileSync(join(__dirname, './__fixtures__/jira.md'), 'utf8');

    expect(stringify(parse(document))).toBe(document);
  });
});
