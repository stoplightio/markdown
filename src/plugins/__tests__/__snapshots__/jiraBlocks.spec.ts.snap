// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jira blocks plugin should parse more complex document correctly 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 22,
              "line": 1,
              "offset": 21,
            },
            "indent": Array [],
            "start": Object {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
          },
          "type": "text",
          "value": "@stoplight/markdown",
        },
      ],
      "depth": 1,
      "position": Position {
        "end": Object {
          "column": 22,
          "line": 1,
          "offset": 21,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "heading",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "alt": "Maintainability",
              "position": Position {
                "end": Object {
                  "column": 96,
                  "line": 3,
                  "offset": 118,
                },
                "indent": Array [],
                "start": Object {
                  "column": 2,
                  "line": 3,
                  "offset": 24,
                },
              },
              "title": null,
              "type": "image",
              "url": "https://api.codeclimate.com/v1/badges/751d2319d7d0fd68d8c9/maintainability",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 166,
              "line": 3,
              "offset": 188,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 3,
              "offset": 23,
            },
          },
          "title": null,
          "type": "link",
          "url": "https://codeclimate.com/github/stoplightio/markdown/maintainability",
        },
        Object {
          "position": Position {
            "end": Object {
              "column": 167,
              "line": 3,
              "offset": 189,
            },
            "indent": Array [],
            "start": Object {
              "column": 166,
              "line": 3,
              "offset": 188,
            },
          },
          "type": "text",
          "value": " ",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 167,
          "line": 3,
          "offset": 189,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 3,
          "offset": 23,
        },
      },
      "type": "paragraph",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 102,
              "line": 5,
              "offset": 292,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 5,
              "offset": 191,
            },
          },
          "type": "text",
          "value": "Useful functions when working with Markdown. Leverages the Unified / Remark ecosystem under the hood.",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 102,
          "line": 5,
          "offset": 292,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 5,
          "offset": 191,
        },
      },
      "type": "paragraph",
    },
    Object {
      "children": Array [
        Object {
          "checked": null,
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "position": Position {
                    "end": Object {
                      "column": 27,
                      "line": 7,
                      "offset": 320,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 3,
                      "line": 7,
                      "offset": 296,
                    },
                  },
                  "type": "text",
                  "value": "Explore the interfaces: ",
                },
                Object {
                  "children": Array [
                    Object {
                      "position": Position {
                        "end": Object {
                          "column": 33,
                          "line": 7,
                          "offset": 326,
                        },
                        "indent": Array [],
                        "start": Object {
                          "column": 28,
                          "line": 7,
                          "offset": 321,
                        },
                      },
                      "type": "text",
                      "value": "TSDoc",
                    },
                  ],
                  "position": Position {
                    "end": Object {
                      "column": 74,
                      "line": 7,
                      "offset": 367,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 27,
                      "line": 7,
                      "offset": 320,
                    },
                  },
                  "title": null,
                  "type": "link",
                  "url": "https://stoplightio.github.io/markdown",
                },
              ],
              "position": Position {
                "end": Object {
                  "column": 74,
                  "line": 7,
                  "offset": 367,
                },
                "indent": Array [],
                "start": Object {
                  "column": 3,
                  "line": 7,
                  "offset": 296,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 74,
              "line": 7,
              "offset": 367,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 7,
              "offset": 294,
            },
          },
          "spread": false,
          "type": "listItem",
        },
        Object {
          "checked": null,
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "position": Position {
                    "end": Object {
                      "column": 23,
                      "line": 8,
                      "offset": 390,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 3,
                      "line": 8,
                      "offset": 370,
                    },
                  },
                  "type": "text",
                  "value": "View the changelog: ",
                },
                Object {
                  "children": Array [
                    Object {
                      "position": Position {
                        "end": Object {
                          "column": 32,
                          "line": 8,
                          "offset": 399,
                        },
                        "indent": Array [],
                        "start": Object {
                          "column": 24,
                          "line": 8,
                          "offset": 391,
                        },
                      },
                      "type": "text",
                      "value": "Releases",
                    },
                  ],
                  "position": Position {
                    "end": Object {
                      "column": 83,
                      "line": 8,
                      "offset": 450,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 23,
                      "line": 8,
                      "offset": 390,
                    },
                  },
                  "title": null,
                  "type": "link",
                  "url": "https://github.com/stoplightio/markdown/releases",
                },
              ],
              "position": Position {
                "end": Object {
                  "column": 83,
                  "line": 8,
                  "offset": 450,
                },
                "indent": Array [],
                "start": Object {
                  "column": 3,
                  "line": 8,
                  "offset": 370,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 83,
              "line": 8,
              "offset": 450,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 8,
              "offset": 368,
            },
          },
          "spread": false,
          "type": "listItem",
        },
      ],
      "ordered": false,
      "position": Position {
        "end": Object {
          "column": 83,
          "line": 8,
          "offset": 450,
        },
        "indent": Array [
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 7,
          "offset": 294,
        },
      },
      "spread": false,
      "start": null,
      "type": "list",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 17,
              "line": 10,
              "offset": 468,
            },
            "indent": Array [],
            "start": Object {
              "column": 5,
              "line": 10,
              "offset": 456,
            },
          },
          "type": "text",
          "value": "Installation",
        },
      ],
      "depth": 3,
      "position": Position {
        "end": Object {
          "column": 17,
          "line": 10,
          "offset": 468,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 10,
          "offset": 452,
        },
      },
      "type": "heading",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 39,
              "line": 12,
              "offset": 508,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 12,
              "offset": 470,
            },
          },
          "type": "text",
          "value": "Supported in modern browsers and node.",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 39,
          "line": 12,
          "offset": 508,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 12,
          "offset": 470,
        },
      },
      "type": "paragraph",
    },
    Object {
      "code": "code",
      "position": Position {
        "end": Object {
          "column": 9,
          "line": 16,
          "offset": 560,
        },
        "indent": Array [
          1,
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 14,
          "offset": 510,
        },
      },
      "type": "jira",
      "value": "yarn add @stoplight/markdown",
    },
    Object {
      "lang": "bash",
      "meta": null,
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 21,
          "offset": 618,
        },
        "indent": Array [
          1,
          1,
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 18,
          "offset": 562,
        },
      },
      "type": "code",
      "value": "# latest stable
yarn add @stoplight/markdown",
    },
    Object {
      "code": "image",
      "position": Position {
        "end": Object {
          "column": 9,
          "line": 27,
          "offset": 661,
        },
        "indent": Array [
          1,
          1,
          1,
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 23,
          "offset": 620,
        },
      },
      "type": "jira",
      "value": "{
  \\"images\\": {}
}",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 1,
      "line": 28,
      "offset": 662,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
